File: bindings/ruby/test/test_diagnostics.rb
require_relative "test_helper"

class TestDiagnostics < Minitest::Test
  def test_severity_values
    doc = Wcl.parse("= invalid")
    assert doc.diagnostics.size > 0
    doc.diagnostics.each do |d|
      assert_includes %w[error warning info hint], d.severity
    end
  end

  def test_error_has_message
    doc = Wcl.parse("= = =")
    assert doc.has_errors?
    doc.errors.each do |e|
      assert_kind_of String, e.message
      assert e.message.size > 0
    end
  end

  def test_valid_doc_no_errors
    doc = Wcl.parse("x = 42")
    refute doc.has_errors?
    assert_equal 0, doc.errors.size
  end

  def test_diagnostic_code
    source = <<~WCL
      schema "cfg" { port: int }
      cfg { port = "bad" }
    WCL
    doc = Wcl.parse(source)
    coded = doc.diagnostics.select { |d| d.code }
    assert coded.size > 0
  end

  def test_diagnostic_inspect
    doc = Wcl.parse("= = =")
    doc.diagnostics.each do |d|
      assert_includes d.inspect, "Diagnostic("
    end
  end

  def test_errors_only_subset
    source = <<~WCL
      @warning
      validation "soft check" {
          let x = -1
          check = x > 0
          message = "x not positive"
      }
    WCL
    doc = Wcl.parse(source)
    doc.errors.each do |e|
      assert_equal "error", e.severity
    end
  end

  def test_diagnostics_include_warnings
    source = <<~WCL
      @warning
      validation "soft check" {
          let x = -1
          check = x > 0
          message = "x not positive"
      }
    WCL
    doc = Wcl.parse(source)
    warnings = doc.diagnostics.select(&:warning?)
    assert warnings.size > 0
  end
end

File: bindings/python/tests/test_diagnostics.py
import wcl


def test_severity_values():
    doc = wcl.parse("= invalid")
    assert len(doc.diagnostics) > 0
    for d in doc.diagnostics:
        assert d.severity in ("error", "warning", "info", "hint")


def test_error_has_message():
    doc = wcl.parse("= = =")
    assert doc.has_errors
    for e in doc.errors:
        assert isinstance(e.message, str)
        assert len(e.message) > 0


def test_valid_doc_no_errors():
    doc = wcl.parse("x = 42")
    assert not doc.has_errors
    assert len(doc.errors) == 0


def test_diagnostic_code():
    source = """
schema "cfg" { port: i64 }
cfg { port = "bad" }
"""
    doc = wcl.parse(source)
    coded = [d for d in doc.diagnostics if d.code is not None]
    assert len(coded) > 0


def test_diagnostic_repr():
    doc = wcl.parse("= = =")
    for d in doc.diagnostics:
        r = repr(d)
        assert "Diagnostic(" in r


def test_errors_only_subset():
    # Warnings should not appear in errors
    source = """
@warning
validation "soft check" {
    let x = -1
    check = x > 0
    message = "x not positive"
}
"""
    doc = wcl.parse(source)
    # The warning validation should produce a warning, not an error
    for e in doc.errors:
        assert e.severity == "error"


def test_diagnostics_include_warnings():
    source = """
@warning
validation "soft check" {
    let x = -1
    check = x > 0
    message = "x not positive"
}
"""
    doc = wcl.parse(source)
    warnings = [d for d in doc.diagnostics if d.severity == "warning"]
    assert len(warnings) > 0
